images/virtualization-artifact/pkg/controller/vm/internal/provisioning.go

```diff
@@ -120,30 +118,102 @@ func (h *ProvisioningHandler) Name() string {
+		return nil
+	case errors.As(err, new(secretNotFoundError)):
+		builder.Status(metav1.ConditionFalse).
+			Reason(vmcondition.ReasonProvisioningNotReady).
+			Message(service.CapitalizeFirstLetter(err.Error()))
+		return nil
+
+	case errors.Is(err, errSecretIsNotValid):
+		builder.Status(metav1.ConditionFalse).
+			Reason(vmcondition.ReasonProvisioningNotReady).
+			Message(fmt.Sprintf("Invalid secret %q: %s", secretKey.String(), err.Error()))
+		return nil
+
+	case errors.As(err, new(unexpectedSecretTypeError)):
 		builder.Status(metav1.ConditionFalse).
 			Reason(vmcondition.ReasonProvisioningNotReady).
-			Message(fmt.Sprintf("Secret %q not found.", secretKey.String()))
+			Message(service.CapitalizeFirstLetter(err.Error()))
 		return nil
+
+	default:
+		return err
+	}
+}
+
+var errSecretIsNotValid = errors.New("secret is not valid")
+
+type secretNotFoundError string
+
+func (e secretNotFoundError) Error() string {
+	return fmt.Sprintf("secret %s not found", string(e))
+}
+
+type unexpectedSecretTypeError string
+
+func (e unexpectedSecretTypeError) Error() string {
+	return fmt.Sprintf("unexpected secret type: %s", string(e))
+}
+
+var cloudInitCheckKeys = []string{
+	"userdata",
+	"userData",
+}
+
+func newProvisioningValidator(reader client.Reader) *provisioningValidator {
+	return &provisioningValidator{
+		reader: reader,
 	}
+}
+
+type provisioningValidator struct {
+	reader client.Reader
+}
+
+func (v provisioningValidator) Validate(ctx context.Context, key types.NamespacedName) error {
+	secret := &corev1.Secret{}
+	err := v.reader.Get(ctx, key, secret)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			return secretNotFoundError(key.String())
+		}
+		return err
+	}
+	switch secret.Type {
+	case virtv2.SecretTypeCloudInit:
+		return v.validateCloudInitSecret(secret)
+	case virtv2.SecretTypeSysprep:
+		return v.validateSysprepSecret(secret)
+	default:
+		return unexpectedSecretTypeError(secret.Type)
+	}
+}
+
+func (v provisioningValidator) validateCloudInitSecret(secret *corev1.Secret) error {
+	if !v.hasOneOfKeys(secret, cloudInitCheckKeys...) {
+		return fmt.Errorf("the secret should have one of data fields %v: %w", cloudInitCheckKeys, errSecretIsNotValid)
+	}
+	return nil
+}
+
+func (v provisioningValidator) validateSysprepSecret(_ *corev1.Secret) error {
+	return nil
+}
+
+func (v provisioningValidator) hasOneOfKeys(secret *corev1.Secret, checkKeys ...string) bool {
 	validate := len(checkKeys) == 0
 	for _, key := range checkKeys {
 		if _, ok := secret.Data[key]; ok {
 			validate = true
 			break
 		}
 	}
-	if !validate {
-		builder.Status(metav1.ConditionFalse).
-			Reason(vmcondition.ReasonProvisioningNotReady).
-			Message(fmt.Sprintf("Secret %q should has one of data fields %v.", checkKeys, secretKey.String()))
-		return nil
-	}
-	builder.Reason(vmcondition.ReasonProvisioningReady).Status(metav1.ConditionTrue)
-	return nil
+	return validate
```

Comment:
Hard to understand without early returns.

Reply:
What do you mean?